File: cpp/Matrix.cpp
template<typename T>
struct Matrix {
    private:
    vector<vector<T>> data;
    int h, w;
    public:
    Matrix(){}
    Matrix(int h, int w, T val=0): data(h, vector<T>(w, val)),h(h), w(w){}

    Matrix unit() const {
        assert(h == w);
        Matrix<T> retu(h, w);
        for(int i = 0;i < h; ++i){
            retu[i][i] = 1;
        }
        return retu;
    }

    const vector<T>& operator[](int i) const {
        return data[i];
    }

    vector<T>& operator[](int i){
        return data[i];
    }

    Matrix operator*(const Matrix& a) const {
        assert(w == a.h);
        Matrix<T> retu(h, a.w);
        for(int i = 0;i < h; ++i){
            for(int j = 0;j < a.w; ++j){
                for(int k = 0;k < w; ++k){
                    retu[i][j] += data[i][k] * a[k][j];
                }
            }
        }
        return retu;
    }

    Matrix pow(long long n) const {
        assert(h == w);
        Matrix retu = Matrix(h,h).unit();
        Matrix x = *this;
        while(n != 0){
            if(n&1){
                retu = retu * x;
            }
            x = x * x;
            n >>= 1;
        }
        return retu;
    }
};

File: rust/Matrix.rs
#[derive(Clone)]
struct Matrix<T> {
    shape: (usize, usize),
    data: Vec<T>,
}

impl<T: Clone + Default> Matrix<T> {
    fn new(h: usize, w: usize) -> Self {
        Self {
            shape: (h, w),
            data: vec![T::default(); w*h],
        }
    }

    fn get(&self, i: usize, j: usize) -> T {
        assert!(i < self.shape.0 && j < self.shape.1);
        self.data[i*self.shape.1 + j].clone()
    }

    fn get_mut(&mut self, i: usize, j: usize) -> &mut T {
        assert!(i < self.shape.0 && j < self.shape.1);
        &mut self.data[i*self.shape.1 + j]
    }

    fn set(&mut self, i: usize, j: usize, data: T) {
        assert!(i < self.shape.0 && j < self.shape.1);
        self.data[i*self.shape.1 + j] = data;
    }
}

impl<T: Clone + Default> From<&Vec<Vec<T>>> for Matrix<T> {
    fn from(data: &Vec<Vec<T>>) -> Self {
        let h = data.len();
        let w = data[0].len();
        assert!(w > 0);
        let mut retu = Matrix {
            shape: (h, w),
            data: vec![T::default(); h*w],
        };
        for i in 0..h {
            for j in 0..w {
                retu.data[i*w + j] = data[i][j].clone();
            }
        }

        retu
    }
}

impl<T: Clone + Default + From<i32>> Matrix<T> {
    fn unit(n: usize) -> Self {
        let mut data = vec![T::default(); n*n];
        for i in 0..n {
            data[i*n + i] = 1.into();
        }

        Self {
            shape: (n, n),
            data,
        }
    }
}

impl<T: Clone + std::ops::AddAssign<T> + std::ops::Mul<T, Output=T> + Default> std::ops::Mul<&Matrix<T>> for &Matrix<T> {
    type Output = Matrix<T>;
    fn mul(self, rhs: &Matrix<T>) -> Matrix<T> {
        assert_eq!(self.shape.1, rhs.shape.0);
        let mut retu = Matrix::new(self.shape.0, rhs.shape.1);
        for i in 0..self.shape.0 {
            for k in 0..self.shape.1 {
                for j in 0..rhs.shape.1 {
                    *retu.get_mut(i, j) += self.get(i, k) * rhs.get(k, j);
                }
            }
        }

        retu
    }
}

impl<T: Clone + std::ops::Mul<T, Output=T> + Default + From<i32> + std::ops::AddAssign>  Matrix<T>{
    fn pow(&self, exp: u64) -> Self {
        assert_eq!(self.shape.0, self.shape.1);

        let mut retu = Self::unit(self.shape.0);
        let mut mul = self.clone();
        let mut exp = exp;
        while exp != 0 {
            if exp%2 == 1 {
                retu = &retu * &mul;
            }
            mul = &mul * &mul;
            exp /= 2;
        }

        retu
    }
}
